storage/store_merge_test.go

```diff
@@ -0,0 +1,241 @@
+		t.Fatal(err)
+	}
+
+	// Verify the merge by looking up keys from both ranges.
+	rangeA := store.LookupRange([]byte("a"), nil)
+	rangeB := store.LookupRange([]byte("c"), nil)
+
+	if !bytes.Equal(rangeA.Desc.StartKey, rangeB.Desc.StartKey) {
+		t.Fatalf("merged range start keys are not equal %q=%q", rangeA.Desc.StartKey, rangeB.Desc.StartKey)
+	}
+	if !bytes.Equal(rangeA.Desc.StartKey, engine.KeyMin) {
+		t.Fatalf("The start key is not equal to KeyMin %q=%q", rangeA.Desc.StartKey, engine.KeyMin)
+	}
+	if !bytes.Equal(rangeA.Desc.EndKey, rangeB.Desc.EndKey) {
+		t.Fatalf("merged range end keys are not equal %q=%q", rangeA.Desc.EndKey, rangeB.Desc.EndKey)
+	}
+	if !bytes.Equal(rangeA.Desc.EndKey, engine.KeyMax) {
+		t.Fatalf("The end key is not equal to KeyMax %q=%q", rangeA.Desc.EndKey, engine.KeyMax)
+	}
+	if rangeA.Desc.RaftID != rangeB.Desc.RaftID {
+		t.Fatalf("merged range ids are not equal %d=%d", rangeA.Desc.RaftID, rangeB.Desc.RaftID)
+	}
+
+	// Try to get values from after the merge.
+	gArgs, gReply = getArgs([]byte("aaa"), rangeA.Desc.RaftID, store.StoreID())
+	if err := store.ExecuteCmd(proto.Get, gArgs, gReply); err != nil ||
+		!bytes.Equal(gReply.Value.Bytes, content) {
+		t.Fatal(err)
+	}
+	gArgs, gReply = getArgs([]byte("ccc"), rangeB.Desc.RaftID, store.StoreID())
+	if err := store.ExecuteCmd(proto.Get, gArgs, gReply); err != nil ||
+		!bytes.Equal(gReply.Value.Bytes, content) {
+		t.Fatal(err)
+	}
+
+	// Put new values after the merge on both sides.
+	pArgs, pReply = putArgs([]byte("aaaa"), content, rangeA.Desc.RaftID, store.StoreID())
+	if err = store.ExecuteCmd(proto.Put, pArgs, pReply); err != nil {
+		t.Fatal(err)
+	}
+	pArgs, pReply = putArgs([]byte("cccc"), content, rangeB.Desc.RaftID, store.StoreID())
+	if err = store.ExecuteCmd(proto.Put, pArgs, pReply); err != nil {
+		t.Fatal(err)
+	}
+
+	// Try to get the newly placed values.
+	gArgs, gReply = getArgs([]byte("aaaa"), rangeA.Desc.RaftID, store.StoreID())
+	if err := store.ExecuteCmd(proto.Get, gArgs, gReply); err != nil || !bytes.Equal(gReply.Value.Bytes, content) {
+		t.Fatal(err)
+	}
+	gArgs, gReply = getArgs([]byte("cccc"), rangeA.Desc.RaftID, store.StoreID())
+	if err := store.ExecuteCmd(proto.Get, gArgs, gReply); err != nil ||
+		!bytes.Equal(gReply.Value.Bytes, content) {
+		t.Fatal(err)
+	}
+}
+
+// TestStoreRangeMergeFirstRange attempts to merge the first range
+// which is illegal.
+func TestStoreRangeMergeFirstRange(t *testing.T) {
+	store := createTestStore(t)
+	defer store.Stop()
+
+	aDesc, _, err := createSplitRanges(store)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Merge the b range back into the a range. This should fail.
+	args, reply := adminMergeArgs(engine.KeyMin, *aDesc, 1, store.StoreID())
+	err = store.ExecuteCmd(proto.AdminMerge, args, reply)
+	if err == nil {
+		t.Fatal("Should not be able to merge the first range")
+	}
+}
+
+// TestStoreRangeMergeDistantRanges attempts to merge two ranges
+// that are not not next to each other.
```

Comment:
Add a TODO to test merging ranges that are not co-located (I assume we don't yet have a good setup for testing that scenario)